src/pwm.rs

```diff
@@ -12,98 +12,136 @@ use core::time::Duration;
+/// # struct Pwm1<S>(S);
+/// # struct Pwm1Pin;
+/// # struct HasPins;
+/// # impl Pwm1<HasPins> {
+/// #     pub fn split(self) -> (Pwm1<()>, [Pwm1Pin; 2]) {
+/// #       (Pwm1(()), [Pwm1Pin, Pwm1Pin])
+/// #     }
+/// # }
+/// # impl Pwm1<()> {
+/// #     pub fn join(self, pins: [Pwm1Pin; 2]) -> Pwm1<HasPins> {
+/// #       Pwm1(HasPins)
+/// #     }
+/// # }
+/// #
+/// # impl <S> hal::pwm::Pwm<u16> for Pwm1<S> {
 /// #     const CHANNELS: usize = 4;
+/// #     const MAX_DUTY: u16 = 1024;
 /// #     type Error = Infallible;
-/// #     type Duty = u16;
+/// #
 /// #     fn try_disable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
 /// #     fn try_enable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
-/// #     fn try_get_duty(&self, _: usize) -> Result<u16, Self::Error> { unimplemented!() }
-/// #     fn try_get_max_duty(&self) -> Result<u16, Self::Error> { Ok(0) }
-/// #     fn try_set_duty(&mut self, _: usize, _: u16) -> Result<(), Self::Error> { Ok(()) }
 /// #     fn try_get_period(&self) -> Result<Duration, Self::Error> { unimplemented!() }
 /// #     fn try_set_period<T>(&mut self, _: T) -> Result<(), Self::Error> where T: Into<Duration> { Ok(()) }
 /// # }
+/// #
+/// # impl hal::pwm::PwmDuty<u16> for Pwm1<HasPins> {
+/// #     fn try_get_duty(&self, _: usize) -> Result<u16, Self::Error> { unimplemented!() }
+/// #     fn try_set_duty<D>(&mut self, _: usize, _: D) -> Result<(), Self::Error> where D: Into<u16> { Ok(()) }
+/// # }
+/// # impl hal::pwm::PwmPin<u16> for Pwm1Pin {
+/// #     const MAX_DUTY: u16 = 1024;
+/// #     type Error = Infallible;
+/// #     fn try_disable(&mut self) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_enable(&mut self) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_get_duty(&self) -> Result<u16, Self::Error> { unimplemented!() }
+/// #     fn try_set_duty<D>(&mut self, _: D) -> Result<(), Self::Error> where D: Into<u16> { Ok(()) }
+/// # }
 /// ```
 // unproven reason: pre-singletons API. The `PwmPin` trait seems more useful because it models independent
 // PWM channels. Here a certain number of channels are multiplexed in a single implementer.
-pub trait Pwm {
+pub trait Pwm<Duty> {
     /// Number of available channels
     const CHANNELS: usize;
 
+    /// Maximum duty cycle value
+    const MAX_DUTY: Duty;
+
     /// Enumeration of `Pwm` errors
     type Error;
 
-    /// Type for the `duty` methods
-    ///
-    /// The implementer is free to choose a float / percentage representation
-    /// (e.g. `0.0 .. 1.0`) or an integer representation (e.g. `0 .. 65535`)
-    type Duty;
-
     /// Disables a PWM `channel`
     fn try_disable(&mut self, channel: usize) -> Result<(), Self::Error>;
 
     /// Enables a PWM `channel`
     fn try_enable(&mut self, channel: usize) -> Result<(), Self::Error>;
 
+    // Note: should you be able to _set_ the period once the channels have been split?
+    // my feeling is, probably not
```

Comment:
i can see this on either side, agreed the key focus is usability but both for hal authors and consumers. imo it'd be pretty unexpected if you split your `PWM` to use in two separate places and one of the altered the operation of other, and should you for some reason have two pins from two `PWM`s passed into a driver the operation of `set_period` is going to be fairly non-obvious.

Reply:
I agree that there are potential problems, and I haven't thought this through yet. I don't know what the right solution would be, and whether there are valid use cases for changing the period of split channels.

I'm saying that embedded-hal should be careful about pre-deciding such questions. Defining these interfaces is already hard enough (because there always seems to be some use case or hardware capability/constraint that wasn't considered). I can totally see adding such rules later, as experience with implementing and using these traits grows (which would be a breaking change, strictly speaking; another reason why I'm concerned about the current push for 1.0).